### PR TITLE
(WIP) Initial support for provisioner communication -- RFC!

### PR DESCRIPTION
**Pull request by sschneid**
Jenkins job would use something like:

```
TERM=velocity shakedown \
  --ssh-key-file ${CLI_TEST_SSH_KEY} \
  --provisioner-url 'http://vmpooler.marathon.mesos:4567' \
  --provisioner-platform ${POOL}
```

### CODE REVIEW

**Comment by sschneid** [comment]
@mesosphere-ci retest this please

File: shakedown/cli/main.py [hunk 1]
```diff
@@ -57,6 +68,42 @@ def cli(**args):
     if not args['no_banner']:
         echo(banner(), n=False)
 
+    if args['provisioner_url']:
+      echo('Provisioning DC/OS cluster...', d='step-maj')
+
+      echo('Checking provisioner URL...', d='step-min', n=False)
+      try:
+        requests.get(args['provisioner_url'])
```
Comment (sschneid) [inline shakedown/cli/main.py]: Would be useful to have a sanity check here -- maybe determine 'provisioner URL type'?)

File: shakedown/cli/main.py [hunk 2]
```diff
@@ -57,6 +68,42 @@ def cli(**args):
...
+        echo(args['provisioner_url'])
+      except:
+        click.secho("error: cannot connect to '{}'.".format(args['provisioner_url']), fg='red', bold=True)
+        sys.exit(1)
+
+      echo('Checking requested cluster platform...', d='step-min', n=False)
+      echo(args['provisioner_platform'])
```
Comment (sschneid) [inline shakedown/cli/main.py]: Needs an actual platform sanity check here.

**Comment by sschneid** [comment]
This a proposal for a way for Shakedown runs to obtain a DC/OS cluster via a 'provisioner service'.  Ideally the `--provisioner-url` flag could point to vmpooler (as shown) or a `dcos launch web` service?

rfc!

File: shakedown/cli/main.py [hunk 3]
```diff
@@ -57,6 +68,42 @@ def cli(**args):
...
+        echo("Waiting for cluster...", d='step-min', n=False)
+
+        response = requests.post("{}/api/v1/vm/{}".format(args['provisioner_url'], args['provisioner_platform'])).json()
+        if ('ok' in response) and (response['ok']):
+          success = True
+          args['cluster_id'] = response[args['provisioner_platform']]['vm']
+          echo(args['cluster_id'])
+          args['dcos_url'] = 'http://' + requests.get("{}/api/v1/vm/{}".format(args['provisioner_url'], args['cluster_id'])).json()[args['cluster_id']]['hostname']
+        else:
+          if attempt < (args['provisioner_max_tries'] + 1):
```
Comment (ssk2) [inline shakedown/cli/main.py]: It looks like this logic will work as long as vmpooler either has or is able to provide a cluster within 60 + 120 + 180 = 360 seconds. Does it make sense to have a longer default wait time and a fixed number of times (with no exponential back off)?

File: shakedown/cli/main.py [hunk 4]
```diff
@@ -57,6 +68,42 @@ def cli(**args):
...
+            sleep = args['provisioner_wait'] * attempt
```
Comment (sschneid) [inline shakedown/cli/main.py]: > I don't think this is an exponential backoff

...okay, true.  :)

File: shakedown/cli/main.py [hunk 5]
```diff
@@ -336,4 +383,15 @@ def pytest_sessionfinish(self, session, exitstatus):
 
     exitstatus = imported['pytest'].main(opts, plugins=[shakedown()])
 
+    if args['provisioner_url']:
+      echo('Cleaning up...', d='step-maj')
+      echo('Destroying cluster...', d='step-min', n=False)
+      response = requests.delete("{}/api/v1/vm/{}".format(args['provisioner_url'], args['cluster_id']))
+      if ('ok' in response) and (response['ok']):
```
Comment (sschneid) [inline shakedown/cli/main.py]: There is retry login on the back-end (vmpooler will retry deleting AWS `FAILED_DELETE` clusters) but making multiple requests won't do anything (the request to the vmpooler API is just saying "move this from `running` to `destroy` queue).

File: shakedown/cli/main.py (hunk 2 above)
Comment (sschneid) [inline shakedown/cli/main.py]: > Could we create a Provisioner class and have a vmpooler implementation?

Yes, I'd like this to be more pluggable / swappable for sure.

File: shakedown/cli/main.py [hunk 6]
```diff
@@ -11,6 +13,10 @@
 @click.command('shakedown')
 @click.argument('tests', nargs=-1)
 @click.option('-u', '--dcos-url', help='URL to a running DC/OS cluster.')
+@click.option('-U', '--provisioner-url', help='URL to a DC/OS provisioning service')
+@click.option('-P', '--provisioner-platform', help='Platform type to request from provisioning service')
+@click.option('--provisioner-wait', default=60, help='Seconds (exponential) to wait between provisioning attempts')
```
Comment (sschneid) [inline shakedown/cli/main.py]: Perhaps the available options flags should be left to the provisioner class.  For vmpooler it's a pre-defined set that can be requested from; for dcos-launch it'd make sense to send along a YAML/JSON payload to specify cluster type/shape.

File: shakedown/cli/main.py (hunk 3 above)
Comment (sschneid) [inline shakedown/cli/main.py]: Yeah, max number of tries and wait time should both be configurable.

**Pull request closed without merging**